File: howdy/src/recorders/video_capture.cpp
#include "recorders/video_capture.hpp"

#include "common/capture_device_path.hpp"
#include "config/config_values.hpp"

#include <filesystem>
#include <string>
#include <utility>

#include <opencv2/imgproc.hpp>

namespace howdy::native {

	namespace {

		constexpr auto kNoDevice = "none";

	}  // namespace

	auto load_capture_settings(const ConfigReader &config) -> CaptureSettings {
		return CaptureSettings{
		    .device_path    = config.get("video", "device_path", "/dev/video0"),
		    .warn_no_device = config.get_bool("video", "warn_no_device", true),
		    .force_mjpeg    = config.get_bool("video", "force_mjpeg", false),
		    .frame_width    = config_frame_width(config),
		    .frame_height   = config_frame_height(config),
		    .device_fps     = config_device_fps(config),
		};
	}

	VideoCapture::VideoCapture(CaptureSettings settings)
	    : settings_(std::move(settings)) {}

	auto VideoCapture::open() -> bool {
		release();
		error_ = CaptureError::kNone;
		error_message_.clear();

		if (settings_.device_path != kNoDevice && !std::filesystem::exists(settings_.device_path)) {
			if (settings_.warn_no_device) {
				set_error(CaptureError::kMissingDevice,
				          "Configured camera device does not exist: " + settings_.device_path);
				return false;
			}
		}

		if (!is_allowed_capture_device_path(settings_.device_path)) {
			set_error(CaptureError::kOpenFailed, "Configured camera device must be a /dev/video* "
			                                     "or /dev/v4l/by-path/* character device: " +
			                                         settings_.device_path);
			return false;
		}

		try {
			capture_.open(settings_.device_path, cv::CAP_V4L);
			if (!capture_.isOpened()) {
				set_error(CaptureError::kOpenFailed,
				          "Failed to open camera device: " + settings_.device_path);
				return false;
			}

			if (settings_.device_fps > 0) {
				capture_.set(cv::CAP_PROP_FPS, settings_.device_fps);
			}
			if (settings_.force_mjpeg) {
				capture_.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));
			}
			if (settings_.frame_width != -1) {
				capture_.set(cv::CAP_PROP_FRAME_WIDTH, settings_.frame_width);
			}
			if (settings_.frame_height != -1) {
				capture_.set(cv::CAP_PROP_FRAME_HEIGHT, settings_.frame_height);
			}

			// Keep a warm-up grab for compatibility with existing behavior; some
			// devices may fail the first grab immediately after open and recover on
			// subsequent reads.
			(void)grab();
			return true;
		} catch (const cv::Exception &error) {
			release();
			set_error(CaptureError::kOpenFailed, "OpenCV failed while opening camera device " +
			                                         settings_.device_path + ": " + error.what());
			return false;
		}
	}

	auto VideoCapture::grab() -> bool {
		if (!capture_.isOpened()) {
			set_error(CaptureError::kOpenFailed, "Camera is not open");
			return false;
		}

		if (!capture_.grab()) {
			set_error(CaptureError::kReadFailed, "Failed to grab a frame from camera");
			return false;
		}
		return true;
	}

	auto VideoCapture::read(cv::Mat &frame, cv::Mat *gray_frame) -> bool {
		if (!capture_.isOpened()) {
			set_error(CaptureError::kOpenFailed, "Camera is not open");
			return false;
		}

		try {
			if (!capture_.read(frame) || frame.empty()) {
				set_error(CaptureError::kReadFailed, "Failed to read a frame from camera");
				return false;
			}
			if (frame.rows <= 0 || frame.cols <= 0) {
				set_error(CaptureError::kReadFailed, "Camera returned invalid frame dimensions");
				return false;
			}

			if (gray_frame != nullptr) {
				if (frame.channels() == 3) {
					cv::cvtColor(frame, *gray_frame, cv::COLOR_BGR2GRAY);
				} else if (frame.channels() == 4) {
					cv::cvtColor(frame, *gray_frame, cv::COLOR_BGRA2GRAY);
				} else if (frame.channels() == 1) {
					*gray_frame = frame;
				} else {
					set_error(CaptureError::kReadFailed,
					          "Camera returned unsupported frame channel count: " +
					              std::to_string(frame.channels()));
					return false;
				}
			}

			return true;
		} catch (const cv::Exception &error) {
			set_error(CaptureError::kReadFailed,
			          "OpenCV failed while reading camera frame: " + std::string(error.what()));
			return false;
		}
	}

	void VideoCapture::release() {
		if (capture_.isOpened()) {
			capture_.release();
		}
	}

	auto VideoCapture::is_open() const -> bool {
		return capture_.isOpened();
	}

	auto VideoCapture::get(int property) const -> double {
		return capture_.isOpened() ? capture_.get(property) : 0.0;
	}

	auto VideoCapture::set(int property, double value) -> bool {
		return capture_.isOpened() && capture_.set(property, value);
	}

	auto VideoCapture::error() const -> CaptureError {
		return error_;
	}

	auto VideoCapture::error_message() const -> const std::string & {
		return error_message_;
	}

	auto VideoCapture::settings() const -> const CaptureSettings & {
		return settings_;
	}

	void VideoCapture::set_error(CaptureError error, std::string message) {
		error_         = error;
		error_message_ = std::move(message);
	}

}  // namespace howdy::native

File: howdy/src/lib/recorders/video_capture.py
# Top level class for a video capture providing simplified API's for common
# functions

from __future__ import annotations

import configparser
import os
import sys
from typing import Optional, Tuple, Union

import cv2
import numpy
from i18n import _

# The internal recorder can be accessed with 'video_capture.internal'


class VideoCapture:
    config: configparser.ConfigParser
    internal: cv2.VideoCapture
    fw: Optional[int]
    fh: Optional[int]
    fps: Optional[int]

    def __init__(self, config: Union[str, configparser.ConfigParser]) -> None:
        """
        Creates a new VideoCapture instance depending on the settings in the
        provided config file.

        Config can either be a string to the path, or a pre-setup configparser.
        """

        if isinstance(config, str):
            self.config = configparser.ConfigParser()
            self.config.read(config)
        else:
            self.config = config

        device_path = self.config.get("video", "device_path")
        if device_path != "none" and not os.path.exists(device_path):
            if self.config.getboolean("video", "warn_no_device", fallback=True):
                print(
                    _(
                        "Howdy could not find a camera device at the path specified in the config file."
                    )
                )
                print(
                    _(
                        "It is very likely that the path is not configured correctly, please edit the 'device_path' config value by running:"
                    )
                )
                print("\n\tsudo howdy config\n")
            sys.exit(14)

        self.internal = None
        self.fw = None
        self.fh = None
        self._create_reader()

        # Request a frame to wake the camera up
        self.internal.grab()

    def __del__(self) -> None:
        try:
            self.internal.release()
        except AttributeError:
            pass  # Internal was never initialized, nothing to release

    def release(self) -> None:
        self.internal.release()

    def read_frame(self) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """
        Reads a frame, returns the frame and an attempted grayscale conversion of
        the frame in a tuple:

        (frame, grayscale_frame)

        If the grayscale conversion fails, both items in the tuple are identical.
        """

        ret, frame = self.internal.read()
        if not ret:
            print(
                _(
                    "Failed to read camera specified in the 'device_path' config option, aborting"
                )
            )
            sys.exit(14)

        try:
            gsframe = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        except RuntimeError:
            gsframe = frame
        except cv2.error:
            print("\nAn error occurred in OpenCV\n")
            raise
        return frame, gsframe

    def _create_reader(self) -> None:
        """
        Sets up the video reader instance
        """
        recording_plugin = self.config.get(
            "video", "recording_plugin", fallback="opencv"
        )
        if recording_plugin != "opencv":
            print(
                _(
                    "Only the OpenCV recorder is supported in this version, forcing recording_plugin=opencv"
                )
            )

        self.internal = cv2.VideoCapture(
            self.config.get("video", "device_path"), cv2.CAP_V4L
        )
        # Without this OpenCV uses the first detected (possibly lower) frame rate.
        # Use 0 as fallback to avoid breaking existing setups.
        self.fps = self.config.getint("video", "device_fps", fallback=0)
        if self.fps != 0:
            self.internal.set(cv2.CAP_PROP_FPS, self.fps)

        if self.config.getboolean("video", "force_mjpeg", fallback=False):
            # Magic number enables MJPEG; badly documented in OpenCV.
            self.internal.set(cv2.CAP_PROP_FOURCC, 1196444237)

        # Set the frame width and height if requested
        self.fw = self.config.getint("video", "frame_width", fallback=-1)
        self.fh = self.config.getint("video", "frame_height", fallback=-1)
        if self.fw != -1:
            self.internal.set(cv2.CAP_PROP_FRAME_WIDTH, self.fw)
        if self.fh != -1:
            self.internal.set(cv2.CAP_PROP_FRAME_HEIGHT, self.fh)
